### PR TITLE
Adding example using ONNX model to get sentence embeddings

### PR DESCRIPTION
**Pull request by SidJain1412**

### CODE REVIEW

**Comment by SidJain1412** [comment]
Hey @nreimers , I would appreciate it if you could review this PR, as I also want some help post merging it to speed the inference up.
Currently the inferences in this notebook take the same time as directly using sentence transformers but I think it should've been faster due to the ONNX optimizations.

**Comment by miketrimmel** [comment]
Would also be very interested in this.

**Comment by miketrimmel** [comment]
@SidJain1412 For me it seems like the mean pooling operation reduces the speed of the procedure to nearly 50% (depending on the sequence length).

I´m actually working on to boost the performance. Will let you know when I get some boost.

**Comment by SidJain1412** [comment]
Awesome @miketrimmel , please keep me updated 😃

**Comment by miketrimmel** [comment]
Did you already find any bottlenecks?

**Comment by SidJain1412** [comment]
@miketrimmel In my testing, the mean pooling operation took very little time, like 1% of the time taken to run the InferenceSession

**Comment by SidJain1412** [comment]
@miketrimmel you can check out [this colab notebook](https://colab.research.google.com/drive/16ChJaw54fSffOVcAoPDeJPDyxGRap8qS?usp=sharing) for my basic experimentation

**Comment by miketrimmel** [comment]
OK, I see. I mean your sequence is very short. I did not run the benchmark with this sequence. I tested it in pre-production with much longer sequences and in a loop, simulating incoming streaming data.

Right now I´m having a mean seq len of about 40 tokens per request. I will have another look at the benchmark today. My latency time is about 10-12ms per request with a quantized distilbert model (distilbert-base-nli-stsb-mean-tokens).

**Comment by fhaase2** [comment]
@SidJain1412 @miketrimmel  it also possible to transform the Pooling layer to onnx format and combine both when doing the inference?

**Comment by SidJain1412** [comment]
Hey @fhaase2 , I'm not really sure if that's possible. If it is, what are the benefits of doing that? I'm not so sure if the pooling process is very time consuming, but maybe @miketrimmel has some benchmark results

**Comment by sidythakur** [comment]
> @miketrimmel you can check out [this colab notebook](https://colab.research.google.com/drive/16ChJaw54fSffOVcAoPDeJPDyxGRap8qS?usp=sharing) for my basic experimentation

I followed the steps mentioned in this colab and tried to create the inferences.
But it throws below error
**onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Invalid Feed Input Name:token_type_ids**

The value of inputs_onnx is:
**{'input_ids': array([[  101,  2026,  2171,  2003, 14324,   102]], dtype=int64), 'token_type_ids': array([[0, 0, 0, 0, 0, 0]], dtype=int64), 'attention_mask': array([[1, 1, 1, 1, 1, 1]], dtype=int64)}**

Can you help me identify the issue here.

Thanks in advance

**Comment by siddhantwaghjale** [comment]
@SidJain1412 Really great work !!
Instead of Sentence -transformers
I'm working on a Spam detection model using fine-tuned DistilBERT model from huggingface.
In order to improve the time-latency in production I'm exporting the finetuned model using the following reference:
 [export transformers to onnx](https://colab.research.google.com/github/huggingface/transformers/blob/master/notebooks/04-onnx-export.ipynb#scrollTo=FDAnSrDZS_vu)

And just replacing the models with my finetuned model's saved directory .
But while predicting ,the normal pytorch output and pytorch+onnx output size are not matching.
Here is a small snipet of the code for prediction:
image

And I was getting the following results:
image

Then I came across your efforts of using pooling,
Can you help me how to interpret the results of pooling for text classification

**Comment by SidJain1412** [comment]
@siddhantwaghjale is the torch output in the example you showed above just the array of 2 floats? Doesn't look right. 
Please have a look at [the notebook](https://github.com/SidJain1412/sentence-transformers/blob/master/examples/onnx/onnx_example.ipynb) and use the `mean_pooling` function to get the same outputs as the one using Torch

**Comment by siddhantwaghjale** [comment]
> @siddhantwaghjale is the torch output in the example you showed above just the array of 2 floats? Doesn't look right.
> Please have a look at [the notebook](https://github.com/SidJain1412/sentence-transformers/blob/master/examples/onnx/onnx_example.ipynb) and use the `mean_pooling` function to get the same outputs as the one using Torch

@SidJain1412  That is the output of direct DistilBERT and it is right as there are only two categories Not-Spam/Spam  as for the tensor output we can directly use softmax to get the classification result.
I just wanted to understand how to use the sequence_embedding output furthur for classification

**Comment by SidJain1412** [comment]
```
import torch

def mean_pooling(model_output, attention_mask):
    model_output = torch.from_numpy(model_output[0])
    token_embeddings = model_output #First element of model_output contains all token embeddings
    attention_mask = torch.from_numpy(attention_mask)
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size())
    sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)
    sum_mask = torch.clamp(input_mask_expanded.sum(1), min=1e-9)
    return sum_embeddings / sum_mask, input_mask_expanded, sum_mask

pooled = cpu_model.run(None, inputs_onnx)
onnx_embeddings = mean_pooling(pooled, inputs_onnx['attention_mask'])[0]

output_onnx = model(onnx_embeddings)
```
This should work, I'm assuming your `model` takes sentence embeddings as input and classifies spam.
`onnx_embeddings` should match your torch embeddings (`torch_input`).

Hope this helps!

**Comment by siddhantwaghjale** [comment]
> ```
> import torch
> 
> def mean_pooling(model_output, attention_mask):
>     model_output = torch.from_numpy(model_output[0])
>     token_embeddings = model_output #First element of model_output contains all token embeddings
>     attention_mask = torch.from_numpy(attention_mask)
>     input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size())
>     sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)
>     sum_mask = torch.clamp(input_mask_expanded.sum(1), min=1e-9)
>     return sum_embeddings / sum_mask, input_mask_expanded, sum_mask
> 
> pooled = cpu_model.run(None, inputs_onnx)
> onnx_embeddings = mean_pooling(pooled, inputs_onnx['attention_mask'])[0]
> 
> output_onnx = model(onnx_embeddings)
> ```
> 
> This should work, I'm assuming your `model` takes sentence embeddings as input and classifies spam.
> `onnx_embeddings` should match your torch embeddings (`torch_input`).
> 
> Hope this helps!

@SidJain1412 There is no other model,In the code the model is DistilBERT
In the snipet the model I used for comparison of normal DistilBERT  with DistilBERT exported in ONNX format,
So is there anyway we can use only the onnx model for classification?
And really appreciating the prompt reply

**Comment by SidJain1412** [comment]
@siddhantwaghjale Is the bert-base-cased.onnx model (in `cpu_model` variable) your classification model in ONNX format?

Can you try 

```
import torch

def mean_pooling(model_output, attention_mask):
    model_output = torch.from_numpy(model_output[0])
    token_embeddings = model_output #First element of model_output contains all token embeddings
    attention_mask = torch.from_numpy(attention_mask)
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size())
    sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)
    sum_mask = torch.clamp(input_mask_expanded.sum(1), min=1e-9)
    return sum_embeddings / sum_mask, input_mask_expanded, sum_mask

pooled = cpu_model.run(None, inputs_onnx)
onnx_embeddings = mean_pooling(pooled, inputs_onnx['attention_mask'])[0]

output_onnx = model(onnx_embeddings)
``` 
and tell me what you get in output_onnx

**Comment by siddhantwaghjale** [comment]
> @siddhantwaghjale Is the bert-base-cased.onnx model (in `cpu_model` variable) your classification model in ONNX format?
> 
> Can you try
> 
> ```
> import torch
> 
> def mean_pooling(model_output, attention_mask):
>     model_output = torch.from_numpy(model_output[0])
>     token_embeddings = model_output #First element of model_output contains all token embeddings
>     attention_mask = torch.from_numpy(attention_mask)
>     input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size())
>     sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)
>     sum_mask = torch.clamp(input_mask_expanded.sum(1), min=1e-9)
>     return sum_embeddings / sum_mask, input_mask_expanded, sum_mask
> 
> pooled = cpu_model.run(None, inputs_onnx)
> onnx_embeddings = mean_pooling(pooled, inputs_onnx['attention_mask'])[0]
> 
> output_onnx = model(onnx_embeddings)
> ```
> 
> and tell me what you get in output_onnx

Yes the 'bert-based-cased.onnx' is a fined-tuned DistilBERT model for classification which is converted in onnx format.
For the code you have given,there is no model I'm using as I'm using the .onnx to replace the model,
The cpu_model_run is the alternative to the model(DistilBertForSequenceClassification) so cannot be used together.
Also for the DistilBertForSequenceClassification we can directly pass the output of tokenizer

**Review by kartikay-eltropy — APPROVED** [review]

**Review by do-me — CHANGES_REQUESTED** [review]
I have two suggestions:

1) I think it would be good to use `AutoTokenizer` here instead to make it easier for other models, e.g. for the current MTEB leaders like `BAAI/bge-base-en-v1.5`. Works also for BERT of course.

```
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("BAAI/bge-base-en-v1.5")
```
2) Also, I think it would be good to make clear that batch processing is possible with this workflow too by simply converting the string to an array and adding items:

```python
# instad of 
query = "My name is Bert"

# an array might be better 
query = ["My name is Bert", "And I am Ernie"]
```
I just tested - no other code changes needd.


Apart from that thank you so much for this notebook - didn't find any working solution elsewhere.
